File: sdk/python/src/cigar_sdk/context_types.py
"""Typed wire values for the local Rust graph (separate from the frozen remote ABI)."""

from typing import Literal, NotRequired, TypedDict

LocalEdgeKind = Literal["requires", "contradicts", "supports", "related"]


class LocalDocument(TypedDict):
    id: str
    source: str
    text: str
    start_line: NotRequired[int]


class LocalContextLimits(TypedDict, total=False):
    max_documents: int
    max_document_bytes: int
    max_total_bytes: int
    max_edges_per_document: int
    max_edges: int
    cache_entries: int
    cache_text_bytes: int


class LocalContextRequest(TypedDict, total=False):
    query: str
    max_tokens: int
    reserve_tokens: int
    required: list[str]
    allowed: list[str] | None
    semantic_candidates: list[str]
    policy_revision: str
    max_candidates: int
    max_blocks: int
    graph_depth: int
    evidence_per_term: int
    excerpt_mode: Literal["full", "query_windows"]


class LocalCitation(TypedDict):
    node_id: str
    source: str
    document_digest: str
    start_line: int
    end_line: int


class LocalContextPrompt(TypedDict):
    schema: Literal["cigar.context-prompt.v1"]
    id: str
    snapshot_id: str
    tokenizer: str
    rendered: str
    rendered_tokens: int
    max_tokens: int
    citations: dict[str, list[LocalCitation]]


class LocalEvidenceBlock(TypedDict):
    text: str
    citations: list[LocalCitation]


class LocalSelectionStats(TypedDict):
    documents: int
    lexical_matches: int
    candidates: int
    selected_blocks: int
    selected_sources: int
    query_terms: int
    covered_terms: int
    rendered_tokens: int
    available_tokens: int
    unavailable_roots: int
    limit_rejected_roots: int
    budget_rejected_roots: int


class LocalContextSnapshot(TypedDict):
    id: str
    domain: str
    policy_revision: str
    tokenizer: str
    graph_revision: int
    blocks: list[LocalEvidenceBlock]
    stats: LocalSelectionStats


class LocalContextResult(TypedDict):
    snapshot: LocalContextSnapshot
    rendered: str


class LocalContextDelta(TypedDict):
    base_id: str
    target_id: str
    graph_revision: int
    stats: LocalSelectionStats
    order: list[str]
    added: list[LocalEvidenceBlock]


class LocalSourceUpdate(TypedDict):
    inserted: int
    replaced: int
    removed: int
    unchanged: int
    revision: int


class LocalTokenCacheStats(TypedDict):
    hits: int
    misses: int
    entries: int
    text_bytes: int


class LocalGraphStats(TypedDict):
    documents: int
    revision: int
    cache: LocalTokenCacheStats

File: sdk/typescript/src/context-types.ts
/** Local graph wire values; these do not replace the frozen remote Context ABI. */
export type LocalEdgeKind = "requires" | "contradicts" | "supports" | "related";
export type LocalDocument = Readonly<{id: string; source: string; text: string; start_line?: number}>;
export type LocalContextLimits = Readonly<{
  max_documents?: number; max_document_bytes?: number; max_total_bytes?: number;
  max_edges_per_document?: number; max_edges?: number; cache_entries?: number; cache_text_bytes?: number;
}>;
export type LocalContextRequest = Readonly<{
  query?: string; max_tokens?: number; reserve_tokens?: number; required?: readonly string[];
  allowed?: readonly string[] | null; semantic_candidates?: readonly string[]; policy_revision?: string;
  max_candidates?: number; max_blocks?: number; graph_depth?: number; evidence_per_term?: number;
  excerpt_mode?: "full" | "query_windows";
}>;
export type LocalCitation = Readonly<{
  node_id: string; source: string; document_digest: string; start_line: number; end_line: number;
}>;
export type LocalEvidenceBlock = Readonly<{text: string; citations: readonly LocalCitation[]}>;
export type LocalSelectionStats = Readonly<{
  documents: number; lexical_matches: number; candidates: number; selected_blocks: number;
  selected_sources: number; query_terms: number; covered_terms: number; rendered_tokens: number;
  available_tokens: number; unavailable_roots: number; limit_rejected_roots: number; budget_rejected_roots: number;
}>;
export type LocalContextSnapshot = Readonly<{
  id: string; domain: string; policy_revision: string; tokenizer: string; graph_revision: number;
  blocks: readonly LocalEvidenceBlock[]; stats: LocalSelectionStats;
}>;
export type LocalContextResult = Readonly<{snapshot: LocalContextSnapshot; rendered: string}>;
export type LocalContextPrompt = Readonly<{
  schema: "cigar.context-prompt.v1"; id: string; snapshot_id: string; tokenizer: string;
  rendered: string; rendered_tokens: number; max_tokens: number;
  citations: Readonly<Record<string, readonly LocalCitation[]>>;
}>;
export type LocalContextDelta = Readonly<{
  base_id: string; target_id: string; graph_revision: number; stats: LocalSelectionStats;
  order: readonly string[]; added: readonly LocalEvidenceBlock[];
}>;
export type LocalSourceUpdate = Readonly<{
  inserted: number; replaced: number; removed: number; unchanged: number; revision: number;
}>;
export type LocalGraphStats = Readonly<{
  documents: number; revision: number;
  cache: Readonly<{hits: number; misses: number; entries: number; text_bytes: number}>;
}>;
